### ISSUE TITLE
the performance of tensorflow distributed training

### ISSUE THREAD

**Issue by dingevin** [mozilla/DeepSpeech#1753]
this is not the performance of DeepSpeech. I have trained a speech recognition network using tensorflow both on multi-gpu and distributed according to your distributed deepspeech. there is some question that i have been  puzzled a long time. that's my [code](https://github.com/dingevin/distributed-training/blob/master/distributed.py), when i'm training i faced two questions:
1. **the accuracy of distributed not equal to multi-gpu**. in multi-gpu version, i use 2 gpu and training 8000 steps, the accuracy can reach 32%; however the accuracy only can reach 28% at distributed version(i have 2 ps 2 worker, and each worker use 2 gpu), each worker training 4000 steps, and the learning rate was same. it seems the parameters not share.
multi-gpu: (when training 4000 steps， loss is reduce to about 5.0)
```
INFO:tensorflow:Step #4015: lr: 0.000100, time 2018-11-28 16:35:16, accuracy 17.12786%, cross entropy 5.230135(147.0 examples/sec; 0.218 sec/batch)
INFO:tensorflow:Step #4016: lr: 0.000100, time 2018-11-28 16:35:17, accuracy 16.93807%, cross entropy 5.064457(151.4 examples/sec; 0.211 sec/batch)
INFO:tensorflow:Step #4017: lr: 0.000100, time 2018-11-28 16:35:17, accuracy 18.51658%, cross entropy 5.146353(149.4 examples/sec; 0.214 sec/batch)
INFO:tensorflow:Step #4018: lr: 0.000100, time 2018-11-28 16:35:17, accuracy 18.98760%, cross entropy 4.905624(145.8 examples/sec; 0.219 sec/batch)
INFO:tensorflow:Step #4019: lr: 0.000100, time 2018-11-28 16:35:17, accuracy 19.00991%, cross entropy 4.929277(139.1 examples/sec; 0.230 sec/batch)
INFO:tensorflow:Step #4020: lr: 0.000100, time 2018-11-28 16:35:17, accuracy 15.84382%, cross entropy 5.140810(139.6 examples/sec; 0.229 sec/batch)
INFO:tensorflow:Step #4021: lr: 0.000100, time 2018-11-28 16:35:18, accuracy 20.37208%, cross entropy 4.693164(143.1 examples/sec; 0.224 sec/batch)
INFO:tensorflow:Step #4022: lr: 0.000100, time 2018-11-28 16:35:18, accuracy 17.22009%, cross entropy 5.180458(140.5 examples/sec; 0.228 sec/batch)
INFO:tensorflow:Step #4023: lr: 0.000100, time 2018-11-28 16:35:18, accuracy 17.57235%, cross entropy 4.930343(142.4 examples/sec; 0.225 sec/batch)
INFO:tensorflow:Step #4024: lr: 0.000100, time 2018-11-28 16:35:18, accuracy 16.51481%, cross entropy 5.369019(140.3 examples/sec; 0.228 sec/batch)
INFO:tensorflow:Step #4025: lr: 0.000100, time 2018-11-28 16:35:19, accuracy 15.62442%, cross entropy 5.515029(138.6 examples/sec; 0.231 sec/batch)
INFO:tensorflow:Step #4026: lr: 0.000100, time 2018-11-28 16:35:19, accuracy 19.32067%, cross entropy 4.870308(141.5 examples/sec; 0.226 sec/batch)
```
distributed ( training 2000 steps， the loss was reduce to 5.5， equal to 2000 steps at multi-gpu  )
```
INFO:tensorflow:Step #2012: learning_rate:0.0001, accuracy 15.16368%, cross entropy 5.732183(19.8 examples/sec; 1.620 sec/batch)
INFO:tensorflow:Step #2013: learning_rate:0.0001, accuracy 15.94258%, cross entropy 5.623287(10.4 examples/sec; 3.077 sec/batch)
INFO:tensorflow:Step #2014: learning_rate:0.0001, accuracy 14.53264%, cross entropy 5.851492(16.1 examples/sec; 1.983 sec/batch)
INFO:tensorflow:Step #2015: learning_rate:0.0001, accuracy 18.74464%, cross entropy 5.227384(18.7 examples/sec; 1.714 sec/batch)
INFO:tensorflow:Step #2016: learning_rate:0.0001, accuracy 15.31315%, cross entropy 5.541482(17.2 examples/sec; 1.862 sec/batch)
INFO:tensorflow:Step #2017: learning_rate:0.0001, accuracy 15.75301%, cross entropy 5.568171(20.7 examples/sec; 1.547 sec/batch)
INFO:tensorflow:Step #2018: learning_rate:0.0001, accuracy 17.56266%, cross entropy 5.382931(15.6 examples/sec; 2.050 sec/batch)
INFO:tensorflow:Step #2019: learning_rate:0.0001, accuracy 16.61609%, cross entropy 5.386977(21.3 examples/sec; 1.505 sec/batch)
INFO:tensorflow:Step #2020: learning_rate:0.0001, accuracy 11.97795%, cross entropy 6.026215(29.3 examples/sec; 1.093 sec/batch)
INFO:tensorflow:Step #2021: learning_rate:0.0001, accuracy 14.76162%, cross entropy 5.657806(32.7 examples/sec; 0.977 sec/batch)
INFO:tensorflow:Step #2022: learning_rate:0.0001, accuracy 15.25826%, cross entropy 5.588447(29.4 examples/sec; 1.087 sec/batch)
INFO:tensorflow:Step #2023: learning_rate:0.0001, accuracy 13.63419%, cross entropy 5.882462(28.7 examples/sec; 1.117 sec/batch)
INFO:tensorflow:Step #2024: learning_rate:0.0001, accuracy 15.66866%, cross entropy 5.585463(37.7 examples/sec; 0.849 sec/batch)
```
2. **the speed of distributed version was so slow**.  the bandwidth is 125Mb/s, is not the bottleneck.

i execute the command as follow on every worker:
```
CUDA_VISIBLE_DEVICES='' nohup python distributed_train.py --ps_hosts=gpu42-hca:2220, gpu47-hca:2221 --worker_hosts=gpu42-hca:3330,gpu47-hca:3331 --job_name=ps --task_index=0 &
CUDA_VISIBLE_DEVICES='0,1' nohup python distributed_train.py --ps_hosts=gpu42-hca:2220,gpu47-hca:2221 --worker_hosts=gpu42-hca:3330,gpu47-hca:3331 --job_name=worker --task_index=0 --gpu_num=2 --learning_rate=0.0001--how_many_training_steps=4000 --save_step_interval=5000 &
```
and another worker:
```
CUDA_VISIBLE_DEVICES='' nohup python distributed_train.py --ps_hosts=gpu42-hca:2220, gpu47-hca:2221 --worker_hosts=gpu42-hca:3330,gpu47-hca:3331 --job_name=ps --task_index=1 &
CUDA_VISIBLE_DEVICES='0,1' nohup python distributed_train.py --ps_hosts=gpu42-hca:2220,gpu47-hca:2221 --worker_hosts=gpu42-hca:3330,gpu47-hca:3331 --job_name=worker --task_index=1 --gpu_num=2 --learning_rate=0.0001--how_many_training_steps=4000 --save_step_interval=5000 &

```
was my code wrong? can you give me some advice. thanks.

**Comment by reuben** [comment]
Sounds like this is better suited for the tensorflow repository? Or at the very least the Discourse forums rather than the issue tracker, since it's not our code.

**Comment by dingevin** [comment]
yes, thanks.
